### PR TITLE
add custom attributes for tchap account configuration in admin

### PR DESCRIPTION
**Pull request by dxb**
With this pr you can specify the tchap account used by the process, directly in client panel of the admin console.

Add client-details template with a new panel for tchap config with :
- is tchap enabled switch
- tchap email account
- tchap password account

Capture d’écran 2022-12-22 à 15 10 44

In code, replace use of  those env variables by new client attributes values.
Fix tests.

### CODE REVIEW

File: src/main/java/org/beta/authentification/keycloak/bot/BotSenderFactory.java [hunk 1]
```diff
@@ -9,9 +9,9 @@
 public class BotSenderFactory {
     private static BotSender instance;
 
-    public static BotSender getInstance() {
+    public static BotSender getInstance(String email, String password) {
```
Comment (odelcroi) [inline src/main/java/org/beta/authentification/keycloak/bot/BotSenderFactory.java:12]: BotSender is a singleton, would it work as expected? Each client should have a different botSender as email/password are per client.

File: src/main/java/org/beta/authentification/keycloak/authenticator/OtpLoginAuthenticatorFactory.java [hunk 2]
```diff
@@ -39,12 +40,24 @@ public class OtpLoginAuthenticatorFactory implements AuthenticatorFactory {
 
     @Override
     public Authenticator create(KeycloakSession session) {
+        String tchapEmail = Environment.getenv(Constants.TCHAP_BOT_ACCOUNT_EMAIL);
```
Comment (odelcroi) [inline src/main/java/org/beta/authentification/keycloak/authenticator/OtpLoginAuthenticatorFactory.java:43]: Constants.TCHAP_BOT_ACCOUNT_EMAIL -> would that be more meaningful to name "TCHAP_BOT_ACCOUNT_EMAIL_DEFAULT" to enhance the fact that it is a default value

File: src/main/java/org/beta/authentification/keycloak/authenticator/OtpLoginAuthenticator.java [hunk 3]
```diff
@@ -104,7 +103,7 @@ public void authenticate(AuthenticationFlowContext context) {
             context.challenge(
                     context.form()
                             .setAttribute(FORM_ATTRIBUTE_USER_EMAIL, user.getUsername())
-                            .setAttribute("feature_tchap_bot", Features.isTchapBotEnabled())
+                            .setAttribute("feature_tchap_bot", Features.isTchapBotEnabled(context.getAuthenticationSession().getClient()))
```
Comment (odelcroi) [inline src/main/java/org/beta/authentification/keycloak/authenticator/OtpLoginAuthenticator.java:106]: can this value be null at any point?

File: src/main/java/org/beta/authentification/keycloak/bot/BotSenderFactory.java (hunk 1 above)
Comment (dxb) [reply src/main/java/org/beta/authentification/keycloak/bot/BotSenderFactory.java:12]: It's working when the singleton is commented.

File: src/main/java/org/beta/authentification/keycloak/authenticator/OtpLoginAuthenticator.java (hunk 3 above)
Comment (dxb) [reply src/main/java/org/beta/authentification/keycloak/authenticator/OtpLoginAuthenticator.java:106]: I have no idea. Maybe there is a case in the process where this value can be null.

File: src/main/java/org/beta/authentification/keycloak/authenticator/OtpLoginAuthenticatorFactory.java (hunk 2 above)
Comment (dxb) [reply src/main/java/org/beta/authentification/keycloak/authenticator/OtpLoginAuthenticatorFactory.java:43]: Definetely, but I think it's less cumbersome for the next release to let it in place like that, in order to reduce friction during first deployement of this new console.

Comment (odelcroi) [reply src/main/java/org/beta/authentification/keycloak/authenticator/OtpLoginAuthenticatorFactory.java:43]: ok

File: src/main/java/org/beta/authentification/keycloak/utils/Features.java [hunk 4]
```diff
@@ -4,9 +4,17 @@
  */
 package org.beta.authentification.keycloak.utils;
 
+import org.apache.commons.lang.StringUtils;
+import org.keycloak.models.ClientModel;
+
 public class Features {
-    public static boolean isTchapBotEnabled() {
-        return Boolean.parseBoolean(Environment.getenv(Constants.FEATURE_TCHAP_BOT_OTP));
+    public static boolean isTchapBotEnabled(ClientModel clientModel) {
```
Comment (odelcroi) [inline src/main/java/org/beta/authentification/keycloak/utils/Features.java:11]: I would rather add a other method with additional parameters than modify the existing code

File: src/main/java/org/beta/authentification/matrix/rest/MatrixServiceFactory.java [hunk 5]
```diff
@@ -34,23 +34,21 @@ private static MatrixService getNonSingletonInstance() {
         return new MatrixService(homeServerList, unauthorizedList);
     }
 
-    public static MatrixService getAuthenticatedInstance() {
+    public static MatrixService getAuthenticatedInstance(String accountEmail, String password) {
```
Comment (odelcroi) [inline src/main/java/org/beta/authentification/matrix/rest/MatrixServiceFactory.java:37]: I would rather add a other method with additional parameters than modify the existing code

File: .env.sample [hunk 6]
```diff
@@ -45,7 +45,7 @@ TCHAP_CODE_TIMEOUT_IN_MINUTES=60
 TCHAP_HOME_REDIRECT_URL=/admin
 
 TCHAP_BOT_ACCOUNT_EMAIL=
-TCHAP_BOT_PASSWORD=""
+TCHAP_BOT_TOKEN=""
```
Comment (odelcroi) [inline .env.sample:48]: N'aura-t-on plus jamais besoin du password?

File: src/main/java/org/beta/authentification/matrix/rest/MatrixServiceFactory.java [hunk 7]
```diff
@@ -54,18 +54,18 @@ private static MatrixService getSingletonAuthenticatedInstance() {
 
     private static MatrixService getNonSingletonAuthenticatedInstance() {
         String accountEmail = Environment.getenv(Constants.TCHAP_BOT_ACCOUNT_EMAIL);
-        String password = Environment.getenv(Constants.TCHAP_BOT_PASSWORD);
+        String token = Environment.getenv(Constants.TCHAP_BOT_TOKEN);
```
Comment (odelcroi) [inline src/main/java/org/beta/authentification/matrix/rest/MatrixServiceFactory.java:57]: ici, on utilise le TCHAP_BOT_TOKEN par défaut? Celui configuré dans les env var? Ca vaut le coup de le renommer desormais je pense

File: .env.sample (hunk 6 above)
Comment (dxb) [reply .env.sample:48]: C'est l'idée, sinon, on risque d'avoir un souci de perf/appel trop nombreux à la mise en prod.

**Review by odelcroi — APPROVED** [review]

**Pull request merged**